File: fixtures/projects/e2e-python/src/helpers/pagination.py
from dataclasses import dataclass

from src.constants.app import DEFAULT_PAGE_SIZE


@dataclass(slots=True)
class Page:
    items: list[dict[str, str]]
    page: int
    page_size: int
    total: int


def paginate(
    items: list[dict[str, str]], page: int, page_size: int = DEFAULT_PAGE_SIZE
) -> Page:
    start = max(0, (page - 1) * page_size)
    end = start + page_size
    return Page(
        items=items[start:end], page=page, page_size=page_size, total=len(items)
    )

File: fixtures/projects/e2e-app/src/helpers/pagination.ts
import { DEFAULT_PAGE_SIZE } from "../constants/app";
import type { PaginationMeta } from "../types/api";

function toPositiveInteger(value: string | undefined, fallback: number): number {
	const parsed = Number.parseInt(value ?? "", 10);
	return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}

export function parsePage(value: string | undefined): number {
	return toPositiveInteger(value, 1);
}

export function parsePageSize(value: string | undefined): number {
	return Math.min(toPositiveInteger(value, DEFAULT_PAGE_SIZE), 100);
}

export function buildPaginationMeta(page: number, pageSize: number, totalItems: number): PaginationMeta {
	const totalPages = Math.max(Math.ceil(totalItems / pageSize), 1);
	return { page, pageSize, totalItems, totalPages };
}